src/http/input/body/N3PatchBodyParser.ts

```diff
@@ -0,0 +1,136 @@
+      throw new BadRequestHttpError(`Invalid N3: ${createErrorMessage(error)}`);
+    }
+    const def = DataFactory.defaultGraph();
+
+    // Solid, §5.3.1: "A patch resource MUST contain a triple ?patch rdf:type solid:InsertDeletePatch."
+    // "The patch document MUST contain exactly one patch resource,
+    // identified by one or more of the triple patterns described above, which all share the same ?patch subject."
+    const subjects = store.getSubjects(RDF.terms.type, SOLID.terms.InsertDeletePatch, def);
+    if (subjects.length !== 1) {
+      throw new UnprocessableEntityHttpError('An N3 Patch document requires exactly 1 solid:InsertDeletePatch entry.');
+    }
+    const subject = subjects[0];
+
+    // Solid, §5.3.1: "A patch resource MUST be identified by a URI or blank node, which we refer to as ?patch
+    // in the remainder of this section."
+    if (subject.termType !== 'NamedNode' && subject.termType !== 'BlankNode') {
+      throw new UnprocessableEntityHttpError('An N3 Patch subject needs to be a blank or named node.');
+    }
+
+    // Extract all quads from the corresponding formulae
+    const deletes = this.findQuads(store, subject, SOLID.terms.deletes, def);
+    const inserts = this.findQuads(store, subject, SOLID.terms.inserts, def);
+    const conditions = this.findQuads(store, subject, SOLID.terms.where, def);
+
+    // Make sure there are no forbidden combinations
+    const conditionVars = this.findVariables(conditions);
+    this.verifyQuads(deletes, conditionVars, 'delete');
+    this.verifyQuads(inserts, conditionVars, 'insert');
+
+    return {
+      deletes,
+      inserts,
+      conditions,
+      binary: true,
+      data: guardedStreamFrom(n3),
+      metadata,
+      isEmpty: false,
+    };
+  }
+
+  /**
+   * Finds all quads in a where/deletes/inserts formula.
+   * The returned quads will be updated so their graph is the default graph instead of the N3 reference to the formula.
+   * Will error in case there are multiple instances of the subject/predicate combination.
+   */
+  private findQuads(store: Store, subject: Quad_Subject, predicate: NamedNode, def: DefaultGraph): Quad[] {
+    const graphs = store.getObjects(subject, predicate, def);
+    if (graphs.length > 1) {
+      throw new UnprocessableEntityHttpError(`An N3 Patch document can have at most 1 ${predicate.value}.`);
+    }
+    if (graphs.length === 0) {
+      return [];
+    }
+    // This might not return all quads in case of nested formulae,
+    // but these are not allowed and will throw an error later when checking for blank nodes.
+    // Another check would be needed in case blank nodes are allowed in the future.
+    const quads: Quad[] = store.getQuads(null, null, null, graphs[0]);
+
+    // Remove the graph references as these are not needed anymore and might influence serializations
+    return quads.map((quad): Quad =>
+      mapTerms(quad, (term, key): Term => key === 'graph' ? def : term));
```

Comment:
I don't understand this.

Reply:
The comment or the line of code?

What we're doing here is removing all the references to the formula the triples were in so they can be interpreted as standard triples. The `N3Patch` object contains 3 sets of triples, so having the graph reference still in there wouldn't make sense. It would make it more difficult to use them as we can't directly apply them to datasets.

Reply:
^ okay, let's add this

I'd personally do a `mapQuads` and return (s, p, o, def), but up to you.